### DISCUSSION TITLE
some questiones about the `Service` implemention of `Router`

### DISCUSSION THREAD

**Discussion by HunterGitHub** [tokio-rs/axum#1968] [Q&A] [answered]
for axum::Router，it implement the `Service` trait
```
impl<B> Service<Request<B>> for Router<(), B>
where
    B: HttpBody + Send + 'static,
{
    type Response = Response;
    type Error = Infallible;
    type Future = RouteFuture<B, Infallible>;

    #[inline]
    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    #[inline]
    fn call(&mut self, req: Request<B>) -> Self::Future {
        self.call_with_state(req, ())
    }
}
```

but the `type Response` is `Response<UnsyncBoxBody<Bytes, Error>>` without generics，maybe it can be changed to `type Response = Response<B>` instead ?  it could support more situations，for example，I want to get `Response<hyper::Body>` using
```
let req: Request<hyper::Body> = ...
let res: Response<hyper::Body> = router.call(req).await
```

**Comment by davidpdrsn** [comment, marked as answer]
> maybe it can be changed to type `Response = Response<B>`

That's not possible because we cannot struct a `B`, so how would we make the response?

```rust
Response::new(???)
```

Therefore the response body type must be type erased so we can combine multiple types that we can construct.

**Reply by HunterGitHub** [reply]
thanks for your answer, I got it

**Reply by HunterGitHub** [reply]
there is some special cases that make me confused:

I start a `hyper::Server` and do some logic using hyper, but if the request info match some special case, I hope to use axum to handle the routing, serde, or some other complex logic, since axum router use `UnsyncBoxBody`, but hyper use its own `hyper::Body`, how can I make it work ?

```
Server::bind(&format!("0.0.0.0:{port}").parse().unwrap())
        .serve(make_service_fn(|_conn| async {
            Ok::<_, Infallible>(service_fn(handler))
        }))

async fn handler(mut req: Request<hyper::Body>) -> Result<Response<hyper::Body>, hyper::Error> {
       if match_some_case(&req) {
             // go axum logic
             return Ok(axum_router.call(req).await.unwrap())
       }

      // go hyper logic
}
```

**Reply by davidpdrsn** [reply]
Use `axum::response::Response` as the response type and create it by calling `.into_resonse()`. Like shown [here](https://docs.rs/axum/latest/axum/response/index.html#returning-different-response-types)

**Reply by HunterGitHub** [reply]
Thanks.
